### ISSUE TITLE
Test collection checks

### ISSUE THREAD

**Issue by yarnabrina** [sktime/sktime#5475]
As of now, there are 3 checks that are being used in different test functions to decide if that should run or not:

1. `_check_soft_dependencies` from `sktime/utils/validation/_dependencies.py`
2. `_check_estimator_deps` from `sktime/utils/validation/_dependencies.py`
3. `run_test_for_class` from `sktime/tests/test_switch.py`

I suggest we create a function `run_test` (or something similar - no preference from me) which is similar to `run_test_for_class` but wihout enforcing `is_class_changed` logic which is part of incremental testing. Then, this can be used for all tests everywhere in `pytest.mark.skipif` decorator, and never use the two private functions any more.

During test collection, let's leave up to the `pytest` configuration to figure out whether to apply increment testing filter or not. If it decides that test is to be run, then always check python and soft-dependency requirements and skip/run accordingly.

CC: @fkiraly

**Comment by fkiraly** [comment]
`run_test_for_class` is meant as the "one-stop-condition" interface that you propose.

Quick question, why would you want to have `is_class_changed` outside?

We should probably get very clear what the condition on running which tests when should be, we seem to have two different sets of ideas here.

We have four types of tests:

A. generic framework tests
B. integration tests or tests certifying for bugfixes
C. `TestAll` suite tests specific to estimator types
D. non-suite tests specific for individual estimators, e.g., logic tests

If an estimator changes, I think we should run:
* C and D specific to the estimator, but not to other estimators
* A and B, but only in the same module

If estimator module specific framework changes, we should run:
* all C and D in the module, but not in other modules
* A and B, full run

If we have one package change in `pyproject`, we should run:
* all C and D for estimators affected by the package change
* all B 

If top framework layer or more serious `pyproject` change, we should run:
* everything

**Comment by fkiraly** [comment]
(labelled this issue as a design discussion - hope that's what you have in mind? Discuss the what, then the implementation plan)

**Comment by yarnabrina** [comment]
In my mental model, I only had A, C and D. I assumed B is implied by A.

Reading your comments, I think it makes sense logically, but not sure if all of these these can be practically achieved or not.

For example, "more serious changes" will probably be difficult to define programmatically even if it's obvious to us. Similarly checking only affected estimators after pyproject gets modified for a single soft dependency ignoring others in same extra is probably impossible, unless we have each dependency in their own extra and have dependencybased folder structure (which was discussed and rejected (?) earlier in #4950). Emphasis on "probably" for both of these though, I may be missing something obvious.

As of now, I realise I don't have a clear testing flow without confusing overlaps. I'll get back by tomorrow.

**Comment by fkiraly** [comment]
> Emphasis on "probably" for both of these though, I may be missing something obvious.

Well, you could use python code to check whether a dependency bound has changed, and tie it to `run_test_for_class`.

**Comment by yarnabrina** [comment]
I am not sure where the different options were being noted from Friday call about extras, igonre, etc. I wanted to suggest one slightly modified option of what I suggested that day.

1. For each "functionality" (e.g. forecasting, classification, etc.), keep a dedicated sub-directory. -> Exists already.
2. For each "functionality", keep a dedicated extra. -> Exists partially.
3. For each "functionality", have unit tests for only that functionality within that sub-directory. -> Exists partially.
4. Keep a dedicated sub-directory for integration tests between different functionalities (e.g. transformation + forecasting, networks + regression, etc.). -> Does not exist (I think?).

Now, let's suppose we have a PR.

1. If it modifies base, it should test everything.
2. If it modifies a subset of functionalities, all unit tests for only the affected functionalities should run independently of one another, and these should install only "functionality" specific dependencies from the dedicated extra's. Also, another independent job should run all integration tests installing all soft dependencies using all_extras.

Proposal Summary: integration tests using multiple functionalities should not be kept under any of those functionality specific sub-directories, and instead those should be moved to a dedicated testing sub-directory.

FYI: @sktime/core-developers 

---

I am aware that one concern was raised against so many "functionality"-specific extra's that new contributors need to know that they need to be edited separately along with all-extras. My argument towards that is pyproject is rarely modified manually as new estimators or new components altogether are not added that frequently.

**Comment by fkiraly** [comment]
> where the different options were being noted from Friday call about extras

I thought the hackmd, thought it seems it either wasn't recorded or has disappeared. Thx for summarizing, will give it a read!

**Comment by yarnabrina** [comment]
@fkiraly @benHeid @achieveordie once the current feature freeze and releases of 0.24.2 and 0.25.0 is done, can we please come back to this? Other than the testing part itself, I am also wondering about the old (test.yml) vs new (test_*.yml and validate_*.yml) CI. Based on my understanding, new one currently covers most of old, and remaining ones are those that IMO not needed to be tested in every PR, and is okay to run every few days:

1. notebooks
2. mlflow
3. cython
4. pandas v1

So, I am hoping we can "retire" old CI in the form of running it every week/month/etc. (schedule + cron) in short term, and in long run disable it with option to trigger manually (workflow_dispatch) as and when needed.

(Not sure if the above question on old vs new CI is more suitable here or in #5101, or to some other issue altogether which I lost track of :( )

**Comment by fkiraly** [comment]
of course! Though is the new CI really running currently? I thought we had not figured out why it is not.

**Comment by yarnabrina** [comment]
I figured out yesterday, it was a typo by referring to a object that doesn't exist any more (renamed the original object but forgot to change references).

Fixed it in #5638 and verified in #5639, ready for review and merge.

**Comment by fkiraly** [comment]
Ahhhh - I see! That hid really well.

I would say, let's merge after 0.25.0, just to avoid any accidents.
